ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java

```diff
@@ -622,3 +630,126 @@ void trackAgeAtAccess(long creationTimeMs) {
     }
   }
 }
+
+/**
+ * A metrics object that is provided as input to {@link CryptoJob}.
+ * </p>
+ * It is expected that each type of crypto job will have it's own instance of CryptoJobMetrics and the same instance is
+ * used to track all jobs of that type.
+ */
+class CryptoJobMetrics {
+  private static final String REQUEST_QUEUING_TIME_SUFFIX = "RequestQueuingTimeInMs";
+  private static final String REQUEST_PROCESSING_TIME_SUFFIX = "RequestProcessingTimeInMs";
+  private static final String RESPONSE_PROCESSING_WAIT_TIME_SUFFIX = "ResponseProcessingWaitTimeInMs";
+  private static final String RESPONSE_PROCESSING_TIME_SUFFIX = "ResponseProcessingTimeInMs";
+  private static final String ROUND_TRIP_TIME_SUFFIX = "RoundTripTimeInMs";
+
+  private static final String OPERATION_RATE_SUFFIX = "Rate";
+  private static final String OPERATION_ERROR_SUFFIX = "Error";
+
+  final Histogram requestQueuingTimeInMs;
+  final Histogram requestProcessingTimeInMs;
+  final Histogram responseProcessingWaitTimeInMs;
+  final Histogram responseProcessingTimeInMs;
+  final Histogram roundTripTimeInMs;
+
+  final Meter operationRate;
+  final Meter operationErrorRate;
+
+  /**
+   * Instantiates {@link CryptoJobMetrics} for {@code requestType} and attaches all the metrics related to the
+   * cryptoJob to the given {@code ownerClass}. The metrics are also registered in the provided {@code metricRegistry}.
+   * @param ownerClass the {@link Class} that is supposed to own the metrics created by this tracker.
+   * @param requestType the type of request for which a tracker is being created.
+   * @param metricRegistry the {@link MetricRegistry} to use to register the created metrics.
+   */
+  CryptoJobMetrics(Class ownerClass, String requestType, MetricRegistry metricRegistry) {
+    requestQueuingTimeInMs =
+        metricRegistry.histogram(MetricRegistry.name(ownerClass, requestType + REQUEST_QUEUING_TIME_SUFFIX));
+    requestProcessingTimeInMs =
+        metricRegistry.histogram(MetricRegistry.name(ownerClass, requestType + REQUEST_PROCESSING_TIME_SUFFIX));
+    responseProcessingWaitTimeInMs =
+        metricRegistry.histogram(MetricRegistry.name(ownerClass, requestType + RESPONSE_PROCESSING_WAIT_TIME_SUFFIX));
+    responseProcessingTimeInMs =
+        metricRegistry.histogram(MetricRegistry.name(ownerClass, requestType + RESPONSE_PROCESSING_TIME_SUFFIX));
+    roundTripTimeInMs = metricRegistry.histogram(MetricRegistry.name(ownerClass, requestType + ROUND_TRIP_TIME_SUFFIX));
+
+    operationRate = metricRegistry.meter(MetricRegistry.name(ownerClass, requestType + OPERATION_RATE_SUFFIX));
+    operationErrorRate = metricRegistry.meter(
+        MetricRegistry.name(ownerClass, requestType + OPERATION_ERROR_SUFFIX + OPERATION_RATE_SUFFIX));
+  }
+}
+
+/**
+ * Construct to support end-to-end metrics tracking for crypto jobs. Usually accompanies a single
```

Comment:
the doc is a little confusing - what do you mean by one-to-one mapping? Multiple instances of these trackers seem to be created in the operations?

Reply:
yes. check out my previous comment.